File: medium/0006_zigzag_conversion/solution.py
class Solution:
    def convert(self, s: str, num_rows: int) -> str:
        if num_rows == 1:
            return s

        s_len = len(s)
        step = 2 * num_rows - 2

        zigzag_s = []

        for row in range(num_rows):
            for i in range(row, s_len, step):
                zigzag_s.append(s[i])

                if (
                    row > 0
                    and row < num_rows - 1
                    and 2 * (num_rows - row - 1) + i < s_len
                ):
                    zigzag_s.append(s[2 * (num_rows - row - 1) + i])

        return "".join(zigzag_s)

File: medium/0006_zigzag_conversion/solution.cpp
#include <string>

class Solution {
public:
  std::string convert(std::string s, int numRows) {
    if (numRows == 1)
      return s;

    const int sLen = s.size();
    const int step = 2 * numRows - 2;

    std::string zigzagS;

    for (int row = 0; row < numRows; ++row) {
      for (int i = row; i < sLen; i += step) {
        zigzagS.push_back(s[i]);

        if (row > 0 && row < numRows - 1 && 2 * (numRows - row - 1) + i < sLen)
          zigzagS.push_back(s[2 * (numRows - row - 1) + i]);
      }
    }

    return zigzagS;
  }
};
